### PR TITLE
improved canceler context

### PR DESCRIPTION
**Pull request by xh3b4sd**
One more time. :D

### CODE REVIEW

File: framework/cancelercontext/canceler_context.go [hunk 1]
```diff
@@ -9,29 +9,33 @@ import (
 // collisions with keys defined in other packages.
 type key string
 
-// cancelKey is the key for cancel values in context.Context. Clients use
+// cancelerKey is the key for canceler values in context.Context. Clients use
```
Comment (xh3b4sd) [inline framework/cancelercontext/canceler_context.go]: This is more consistent now.

File: framework/cancelercontext/canceler_context.go [hunk 2]
```diff
@@ -9,29 +9,33 @@ import (
...
 // key directly.
-var cancelKey key = "canceler"
+var cancelerKey key = "canceler"
 
 // NewContext returns a new context.Context that carries value v.
 func NewContext(ctx context.Context, v chan struct{}) context.Context {
-	return context.WithValue(ctx, cancelKey, v)
+	if v == nil {
+		return ctx
+	}
```
Comment (xh3b4sd) [inline framework/cancelercontext/canceler_context.go]: Potential bug fixed due to testing.

File: framework/cancelercontext/canceler_context.go [hunk 3]
```diff
@@ -9,29 +9,33 @@ import (
...
 	return v, ok
 }
 
 // IsCanceled checks whether the given context obtains information about the
 // canceler as defined in this package, if any canceler is present.
 //
 // NOTE that the canceler, if any found, is only used to be closed to signal
-// cancelatin. It is not guaranteed that the channel is buffered or read from.
-// Clients must not write to it. Otherwise the cancler will block eventually.
-// received.
```
Comment (xh3b4sd) [inline framework/cancelercontext/canceler_context.go]: Whatever happened there. :D

File: framework/framework_test.go [hunk 4]
```diff
@@ -504,9 +504,8 @@ func (r *testResource) GetCurrentState(ctx context.Context, obj interface{}) (in
 	r.Order = append(r.Order, m)
 
 	if r.CancelingStep == m {
-		canceler, exists := cancelercontext.FromContext(ctx)
-		if exists {
-			close(canceler)
+		cancelercontext.SetCanceled(ctx)
+		if cancelercontext.IsCanceled(ctx) {
```
Comment (xh3b4sd) [inline framework/framework_test.go]: It is more convenient now from the client side.

**Comment by xh3b4sd** [comment]
Sorry the diff is pretty shite now.

File: framework/canceledcontext/canceled_context.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+// NewContext returns a new context.Context that carries value v.
+func NewContext(ctx context.Context, v chan struct{}) context.Context {
+	if v == nil {
+		return ctx
+	}
+
+	return context.WithValue(ctx, canceledKey, v)
+}
+
+// FromContext returns the canceled, if any.
```
Comment (rossf7) [inline framework/canceledcontext/canceled_context.go]: "the canceled context" reads better to me.

File: framework/canceledcontext/canceled_context.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+func FromContext(ctx context.Context) (chan struct{}, bool) {
+	v, ok := ctx.Value(canceledKey).(chan struct{})
+	return v, ok
+}
+
+// IsCanceled checks whether the given context obtains information about
+// the canceled as defined in this package, if any canceled is
```
Comment (rossf7) [inline framework/canceledcontext/canceled_context.go]: "canceled context"

File: framework/canceledcontext/canceled_context.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// present.
+//
+// NOTE that the canceled channel, if any found, is only used to be closed to
+// signal cancelation. It is not guaranteed that the channel is buffered or read
+// from. Clients must not write to it. Otherwise the canceled channel will block
+// eventually. It is save to signal cancelation via SetCanceled.
```
Comment (rossf7) [inline framework/canceledcontext/canceled_context.go]: "safe to signal"

File: framework/canceledcontext/canceled_context.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+			return true
+		default:
+			// fall thorugh
+		}
+	}
+
+	return false
+}
+
+// SetCanceled is a save way to signal cancelation.
```
Comment (rossf7) [inline framework/canceledcontext/canceled_context.go]: "safe"

File: framework/canceledcontext/canceled_context.go (hunk 6 above)
Comment (xh3b4sd) [reply framework/canceledcontext/canceled_context.go]: Done.

File: framework/canceledcontext/canceled_context.go (hunk 7 above)
Comment (xh3b4sd) [reply framework/canceledcontext/canceled_context.go]: Done.

File: framework/canceledcontext/canceled_context.go (hunk 5 above)
Comment (xh3b4sd) [reply framework/canceledcontext/canceled_context.go]: It is actually the channel that is returned. But thanks for the hint. Done.

File: framework/canceledcontext/canceled_context.go (hunk 8 above)
Comment (xh3b4sd) [reply framework/canceledcontext/canceled_context.go]: Done.

**Pull request merged**